module/zfs/zfs_vnops.c

```diff
@@ -4922,6 +4923,736 @@ convoff(struct inode *ip, flock64_t *lckdat, int  whence, offset_t offset)
 	return (0);
 }
 
+/*
+ * Convert the provided block pointer in to an extent.  This may result in
+ * a new extent being created or an existing extent being extended.
+ */
+static int
+zfs_fiemap_cb(spa_t *spa, zilog_t *zilog, const blkptr_t *bp,
+    const zbookmark_phys_t *zb, const dnode_phys_t *dnp, void *arg)
+{
+	zfs_fiemap_t *fm = (zfs_fiemap_t *)arg;
+	blkptr_t bp_copy = *bp;
+
+	if (BP_GET_LEVEL(bp) != 0)
+		return (0);
+
+	/*
+	 * Indirect block pointers must be remapped to reflect the real
+	 * physical offset and length.  The remapping is transparent to
+	 * the fiemap interface so no additional extent flags are set.
+	 */
+	spa_config_enter(spa, SCL_VDEV, FTAG, RW_READER);
+	if (spa_remap_blkptr(spa, &bp_copy, NULL, NULL))
+		bp = &bp_copy;
```

Comment:
What if it couldn't be remapped (e.g. has split blocks)?  Then you'll report the offset in the indirect vdev?  I think we should use vdev_op_remap() to report all the segments (I guess this is how you could get a sub-DMU-block-size entry).  If that doesn't work then we could fall back on reporting these segments as FIEMAP_EXTENT_UNKNOWN (like gang blocks).

Reply:
I'll see about reporting all the segments, if that doesn't work I'll set `FIEMAP_EXTENT_UNKNOWN`.